contracts/reverseRegistrar/L2ReverseRegistrar.sol

```diff
@@ -0,0 +1,483 @@
+
+error InvalidSignature();
+
+contract L2ReverseRegistrar is
+    Multicallable,
+    Ownable,
+    ITextResolver,
+    INameResolver,
+    IL2ReverseRegistrar
+{
+    using ECDSA for bytes32;
+    mapping(bytes32 => uint256) public lastUpdated;
+    mapping(uint64 => mapping(bytes32 => mapping(string => string))) versionable_texts;
+    mapping(uint64 => mapping(bytes32 => string)) versionable_names;
+    mapping(bytes32 => uint64) internal recordVersions;
+    event VersionChanged(bytes32 indexed node, uint64 newVersion);
+    bytes32 public immutable L2_REVERSE_NODE;
+
+    bytes32 constant lookup =
+        0x3031323334353637383961626364656600000000000000000000000000000000;
+
+    event ReverseClaimed(address indexed addr, bytes32 indexed node);
+
+    /**
+     * @dev Constructor
+     */
+    constructor(bytes32 L2ReverseNode) {
+        L2_REVERSE_NODE = L2ReverseNode;
+    }
+
+    modifier authorised(address addr) {
+        isAuthorised(addr);
+        _;
+    }
+
+    modifier authorisedSignature(
+        bytes32 hash,
+        address addr,
+        uint256 inceptionDate,
+        bytes memory signature
+    ) {
+        isAuthorisedWithSignature(hash, addr, inceptionDate, signature);
+        _;
+    }
+
+    function isAuthorised(address addr) internal view returns (bool) {
+        require(
+            addr == msg.sender || ownsContract(addr, msg.sender),
+            "ReverseRegistrar: Caller is not a controller or authorised by address or the address itself"
+        );
+    }
+
+    function isAuthorisedWithSignature(
+        bytes32 hash,
+        address addr,
+        uint256 inceptionDate,
+        bytes memory signature
+    ) internal view returns (bool) {
+        bytes32 message = hash.toEthSignedMessageHash();
+        bytes32 node = _getNamehash(addr);
+
+        if (
+            !SignatureChecker.isValidSignatureNow(addr, message, signature) ||
+            inceptionDate < lastUpdated[node] || // must be newer than current record
+            inceptionDate >= block.timestamp // must be in the past
+        ) {
+            revert InvalidSignature();
+        }
+    }
+
+    /**
+     * @dev Sets the name for an addr using a signature that can be verified with ERC1271.
+     * @param addr The reverse record to set
+     * @param name The name of the reverse record
+     * @param inceptionDate Date from when this signature is valid from
+     * @param signature The resolver of the reverse node
+     * @return The ENS node hash of the reverse record.
+     */
+    function setNameForAddrWithSignature(
+        address addr,
+        string memory name,
+        uint256 inceptionDate,
+        bytes memory signature
+    )
+        public
+        override
+        authorisedSignature(
+            keccak256(
+                abi.encodePacked(
+                    IL2ReverseRegistrar.setNameForAddrWithSignature.selector,
+                    addr,
+                    name,
+                    inceptionDate
```

Comment:
How could we do this? I thought about just passing through parameters and hashing inside the modifier, but it means I'd need a modifier per signature, because the signatures differ slightly per function